OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs

```diff
@@ -161,32 +147,39 @@ void BuildUnit(IBot bot, string name)
 
 		ActorInfo ChooseRandomUnitToBuild(ProductionQueue queue)
 		{
-			var buildableThings = queue.BuildableItems();
-			if (!buildableThings.Any())
+			var unitsToBuild = Info.UnitsToBuild.ToArray();
+			if (unitsToBuild.Length == 0)
 				return null;
 
-			var unit = buildableThings.Random(world.LocalRandom);
-			return HasAdequateAirUnitReloadBuildings(unit) ? unit : null;
-		}
-
-		ActorInfo ChooseUnitToBuild(ProductionQueue queue)
-		{
-			var buildableThings = queue.BuildableItems();
-			if (!buildableThings.Any())
+			var buildableThings = queue.BuildableItems().Shuffle(world.LocalRandom).ToArray();
+			if (buildableThings.Length == 0)
 				return null;
 
-			var myUnits = player.World
-				.ActorsHavingTrait<IPositionable>()
-				.Where(a => a.Owner == player)
-				.Select(a => a.Info.Name).ToList();
+			var allUnits = world.Actors.Where(a => a.Owner == player && Info.UnitsToBuild.ContainsKey(a.Info.Name) && !a.IsDead && a.IsInWorld).ToArray();
```

Comment:
Yes. This seems expensive.

Reply:
Well, at least it is better than first use old `BuildUnit` then check the unit limit on line 119 to line 131, which cause two search on the world actors.

Reply:
```suggestion
			var allUnits = world.Actors.Where(a => a.Owner == player && Info.UnitsToBuild.ContainsKey(a.Info.Name) && !a.IsDead).ToArray();
```

we should count all actors no?

Reply:
Idk, but we can do that.

Reply:
IsInWorld check removed